src/plugins/explore/public/application/utils/state_management/middleware/query_sync_middleware.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Middleware } from '@reduxjs/toolkit';
+import { isEqual } from 'lodash';
+import { RootState } from '../store';
+import { ExploreServices } from '../../../../types';
+import { getQueryWithSource } from '../../languages';
+
+/**
+ * Middleware to sync Redux query state with global queryStringManager
+ * This ensures that whenever query state changes, the global state is updated
+ * before any search operations occur. It also manages query history for
+ * user-initiated query executions.
+ */
+export const createQuerySyncMiddleware = (services: ExploreServices): Middleware<{}, RootState> => {
+  return (store) => (next) => (action) => {
+    const result = next(action);
+
+    if (
+      action.type === 'query/setQueryState' ||
+      action.type === 'query/setQueryWithHistory' ||
+      action.type === 'query/setQueryStringWithHistory'
+    ) {
+      const state = store.getState();
+      const query = state.query;
+      const queryWithSource = getQueryWithSource(query);
+
+      if (query.dataset && services.data?.query?.queryString) {
+        const queryStringQuery = services.data.query.queryString.getQuery();
+
+        if (!isEqual(queryStringQuery, queryWithSource)) {
+          services.data.query.queryString.setQuery(queryWithSource);
+        }
+
+        // Add to query history only for user-initiated query executions
+        // This prevents programmatic updates (loading saved queries, clearing, etc.) from polluting history
+        if (action.meta?.addToHistory && query.query?.trim()) {
+          const timefilter = services?.data?.query?.timefilter?.timefilter;
```

Comment:
on line 32 we access services.data.query without the optional chaining. why do we do it here?